File: data structures and algorithms/python/08_008.py
def longest_palindrome(s: str) -> str:

    if not s:
        return ""
    
    start = 0
    max_length = 1
    
    # Helper function to expand around center
    def expand_around_center(left, right):
        nonlocal start, max_length
        while (left >= 0 and right < len(s) and 
               s[left] == s[right]):
            current_length = right - left + 1
            if current_length > max_length:
                max_length = current_length
                start = left
            left -= 1
            right += 1
    
    # Check each position as potential center
    for i in range(len(s)):
        # Odd length palindromes (like "aba")
        expand_around_center(i, i)
        # Even length palindromes (like "abba")
        expand_around_center(i, i + 1)
    
    return s[start:start + max_length]

# Example usage
print(longest_palindrome("babad"))  # "bab" or "aba"
print(longest_palindrome("cbbd"))   # "bb"

File: data structures and algorithms/cpp/08_008.cpp
#include <iostream>
#include <string>

std::string longestPalindrome(std::string s) {
    if (s.empty()) return "";
    
    int start = 0;
    int max_length = 1;
    
    // Helper function to expand around center
    auto expandAroundCenter = [&](int left, int right) {
        while (left >= 0 && right < s.size() && 
               s[left] == s[right]) {
            int current_length = right - left + 1;
            if (current_length > max_length) {
                max_length = current_length;
                start = left;
            }
            left--;
            right++;
        }
    };
    
    // Check each position as potential center
    for (int i = 0; i < s.size(); i++) {
        // Odd length palindromes (like "aba")
        expandAroundCenter(i, i);
        // Even length palindromes (like "abba")
        expandAroundCenter(i, i + 1);
    }
    
    return s.substr(start, max_length);
}

int main() {
    std::cout << longestPalindrome("babad") << std::endl;  
                                            // "bab" or "aba"
    std::cout << longestPalindrome("cbbd") << std::endl;   
                                            // "bb"
    return 0;
}